src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import random
+from enum import Enum, auto
+from itertools import cycle
+
+import torch
+import torch.distributed as dist
+import torch.distributed.distributed_c10d as c10d
+from torch._C._distributed_c10d import ProcessGroup
+from torch.distributed.algorithms._comm_hooks import default
+
+
+class Topology(Enum):
+    r"""
+    Specifies which topology will be used as a base for gradient communication.
+    For more information, please refer to the original
+    `paper <https://arxiv.org/abs/1803.05880>`_
+
+    CUBE: A hypercube topology - a hierarchical virtual organization of compute nodes.
+        For this topology gossiping is happening with a neighboring vertex.
+              *----*
+             /|   /|
+            *----* |
+            | * -|-*
+            |/   |/
+            *----*
+
+    DISSEMINATION: A dissemination topology has similar property
+        as hypercube virtual topology.
+        For this topology gossiping is happening with the neighboring node,
+        then every 2nd node, every 4th, etc.
+
+            .  *  .
+          *          *
+        .              .
+        *              *
+        .              .
+          *          *
+            .  *  .
```

Comment:
curious, do these graphs render properly? :)

Reply:
No, but hereis a screenshot of how they look in an updated version, which I will push soon
<img width="891" alt="Screen Shot 2022-07-27 at 11 17 18 AM" src="https://user-images.githubusercontent.com/13810604/181344109-5c6debd7-96e5-4c0e-b0a7-4c5979ae3db6.png">